util/cache/client.go

```diff
@@ -10,18 +10,35 @@ var ErrCacheMiss = errors.New("cache: key is missing")
 var ErrCacheKeyLocked = errors.New("cache: key is locked")
 var CacheLockedValue = "locked"
 
-type Item struct {
-	Key    string
-	Object interface{}
+type CacheType int
+
+const (
+	CacheTypeDefault CacheType = iota
+	CacheTypeInMemory
+	CacheTypeTwoLevel
+	CacheTypeExternal
+)
+
+type CacheActionOpts struct {
+	// Delete item from cache
+	Delete bool
 	// Disable writing if key already exists (NX)
 	DisableOverwrite bool
 	// Expiration is the cache expiration time.
 	Expiration time.Duration
+	// Determines the cache type to use in case of two level cache, Default == Two Level(in-memory + external)
+	CacheType CacheType
+}
+
+type Item struct {
+	Key    string
+	Object interface{}
+	CacheActionOpts
 }
```

Comment:
Any strong reason to implement this struct as part of this PR? It changes the signature of the `CacheClient` interface. It seems an interesting refactoring but I think it would be better to keep this PR as small as possible. How about provide code refactoring only in a separate PR? The reason is that it allows us to easily rollback changes if required.

Reply:
The primary reason is for (2) in my description here. The need for `DisableOverwrite` (redis NX or memcache add) otherwise depending on timing each caller could think they were the lock holder, of course in practice that likely wouldn't be the case but it would cause inconsistent behavior.

To add that as an option I'd have to modify the caller anyway to add another option.

This struct already exists as well it's in line 13 of the diff. I just add `CacheActionOpts` to it.